homework-submission/week2/main.js

```diff
@@ -0,0 +1,171 @@
+    description: 'Brush teeth',
+    deadline: '2017-09-10',
+    done: false
+  },
+  {
+    id: 3,
+    description: 'Eat breakfast',
+    deadline: '2017-09-09',
+    done: false
+  }
+];
+
+// Material Icon component
+const Icon = ({ children }) => <i className="material-icons">{children}</i>;
+
+// Task item component
+const TaskItem = ({ task, onRemove, onToggleMarked }) => (
+  <StyledListItem>
+    <div onClick={() => onToggleMarked(task.id)} className={task.done ? 'done' : ''}>
+      {task.done && <Icon>check</Icon>}
+      <StyledText bold>{task.description}</StyledText>
+      <StyledText small>{task.deadline}</StyledText>
+    </div>
+    <div>
+      <StyledButton inverse theme="secondary" onClick={() => onRemove(task.id)}>
+        <Icon>delete</Icon>
+      </StyledButton>
+    </div>
+  </StyledListItem>
+);
+
+// Task list component
+const TaskList = props => {
+  const sortedTasks = props.tasks.sort((a, b) => new Date(a.deadline) - new Date(b.deadline));
+  const totalTasks = props.tasks.length;
+  const totalCompletedTasks = sortedTasks.filter(task => task.done).length;
+  const unCompletedTasks = props.tasks.filter(task => !task.done);
+  const completedTasks = props.tasks.filter(task => task.done);
+
+  return (
+    <React.Fragment>
+      <StyledTitle>
+        Sorted Task List
+        <StyledText small>({totalCompletedTasks + '/' + totalTasks})</StyledText>
+      </StyledTitle>
+      <nav>
+        {completedTasks.map(task => (
+          <TaskItem key={task.id} task={task} {...props} />
+        ))}
+        {unCompletedTasks.map(task => (
+          <TaskItem key={task.id} task={task} {...props} />
+        ))}
+      </nav>
+    </React.Fragment>
+  );
+};
+
+// Task add form component
+const TaskAddForm = ({ onAdd }) => {
+  const handleSubmit = e => {
+    e.preventDefault();
+    const description = e.target.description.value;
+    const deadline = e.target.deadline.value;
+
+    onAdd(description, deadline);
+  };
+
+  // Source: https://css-tricks.com/prefilling-date-input/
+  const today = new Date().toISOString().substr(0, 10);
+
+  return (
+    <form name="form" onSubmit={handleSubmit}>
+      <StyledInput name="description" type="text" placeholder="Description" />
+      <StyledInput name="deadline" type="text" placeholder="Deadline" onFocus={e => (e.target.value = today)} />
+      <StyledButton full theme="primary">
+        <Icon>add</Icon>
+      </StyledButton>
+    </form>
+  );
+};
+
+// Task container component
+class TaskContainer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      tasks: this.props.data,
+      lastId: this.props.data.length
+    };
+  }
+
+  handleAdd(description, deadline) {
+    // I any field is empty, alert user without adding the task
+    if (description === '' || deadline === '') return alert('Please fill all the fields!');
+
+    const tasks = [...this.state.tasks];
+    let _lastId = this.state.lastId;
+    tasks.push({
```

Comment:
although this is completely fine
prefer not mutating (pushing to an array) but make a new array

 const tasks = [...this.state.tasks, {
     id: lastId + 1,
      description,
      deadline,
      done: false
}];